3.0.0 rc erikvk

erikvk opened the pull request:
I have finished the documentation for the commands and sub-commands of the star tool, as provided by @per-samuelsson. The goal of this article was to list the commands and provide some basic reference for how to use them, but acknowledging that most people reading it will be familiar with CLI tools already.

- Please review the changes to the `star-tool.md` file
- I have used [Per's document](https://starcounter.slack.com/files/U5H2VB1QX/FRR977ZNH/cli_help.txt) as the SSOT for what should be included in the article.
- Let me know if you feel that we should provide more information or examples on any of the commands.

erikvk committed:
Minor changes, added foundation for star tool docs

erikvk committed:
Merge branch '3.0.0-rc' into 3.0.0-rc-erikvk

erikvk committed:
Finished documenting the commands and subcommands of the star tool.

erikvk committed:
Fixed link to downloads page

erikvk committed:
Learned how to count

erikvk committed:
simplified a sentence

erikvk committed:
minor fixes

erikvk committed:
more minor fixes

Commit:
Merge branch '3.0.0-rc' into 3.0.0-rc-erikvk

Commit:
Update star-tool.md

Commit:
Update star-tool.md

miyconst commented on docs/star-tool.md:
```diff
@@ -0,0 +1,224 @@
+- `i` – For each column, whether it is inherited.
+- `b` – The name of the table of the column.
+
+To list all user columns of the `MyApp.Superhero` table of an existing database located at `C:\databases\mydb`, including inherited columns and listing only the the names and the datatypes of columns, we run:
```
```suggestion
To list all user columns of the `MyApp.Superhero` table of an existing database located at `C:\databases\mydb`, including inherited columns and listing only the names and the datatypes of columns, we run:
```

Commit:
Update docs/star-tool.md

miyconst commented on docs/star-tool.md:
```diff
@@ -0,0 +1,226 @@
+
+Formatting options:
+
+- `c` – The name of the column that the index is registered on.
```
```suggestion
- `c` – The name of the columns that the index is registered on.
```

Commit:
Update docs/star-tool.md

miyconst commented on docs/star-tool.md:
```diff
@@ -0,0 +1,226 @@
+
+Formatting options:
+
+- `c` – The name of the columns that the index is registered on.
```
```suggestion
- `c` – The name of the column that the index is registered on.
```

Commit:
Update docs/star-tool.md

miyconst commented on docs/star-tool.md:
```diff
@@ -0,0 +1,226 @@
+
+Formatting options:
+
+- `c` – The name of the column that the index is registered on.
```
@per-samuelsson how does the tool treat multi-column indexes? And how does it treat when the columns belong to different tables?

per-samuelsson replied:
> @per-samuelsson how does the tool treat multi-column indexes? 

It writes out one column per line. So if you have a multi-column index Foo (Bar, Fubar), you should see:

```
info index -f c Foo
Bar
Fubar
```

> And how does it treat when the columns belong to different tables?

AFAIK, we don't support that. But it will print the name of the table if you tell it so. Relates to the question I asked on Slack yesterday: https://starcounter.slack.com/archives/C5Q4Y78F5/p1576664501013700

bigwad replied:
>AFAIK, we don't support that

Actually, we do (btw, it would be another item to think about if we decide to implement multi-column index attributes ).

per-samuelsson replied:
> ... btw, it would be another item to think about if we decide to implement multi-column index attributes

Please don't mix topics now. 😃

miyconst commented:
@erikvk thank a lot for such a nice page! I have made some adjustments, please review and merge.

per-samuelsson committed:
Typo, expore -> explore

per-samuelsson commented on docs/star-tool.md:
````diff
@@ -0,0 +1,226 @@
+
+The `star` tool is a console application with a command-line interface, and we interact with it using commands, options and arguments.
+
+The global command pattern is:
+
+```
+star [options] [command]
````
I find this inconsistent with all the concrete examples. For example, just a few lines later, we have this:

```
star new [options] <path>
```

where options come after the command.

erikvk replied:
Well, then you should change the tool :)

per-samuelsson commented on docs/star-tool.md:
```diff
@@ -0,0 +1,226 @@
+
+#### Import and export
+
+##### `reload`
```
What about having `unload` before `reload`? That's the normal flow, and the later is dependent on the former.

erikvk replied:
In this case, the reason is trivial: they’re “import and export” commands, hence first reload then unload. But I guess the order is not that important, I can change it

per-samuelsson replied:
> In this case, the reason is trivial: they’re “import and export” commands, hence first reload then unload. But I guess the order is not that important, I can change it

Just figured it's hard to import (reload) something you haven't exported (unloaded) yet. Especially so since we state that we can only ever import data files exported by us. 😎 

But not important, do as you want.

per-samuelsson commented:
> @erikvk thank a lot for such a nice page! I have made some adjustments, please review and merge.

Please don't merge until I've reviewed it. I have comments already.

per-samuelsson reviewed (commented):
A starting review. I'll continue reading, but feel free to merge as you see fit, and then we can adjust later if necessary.

per-samuelsson commented on docs/star-tool.md:
```diff
@@ -0,0 +1,226 @@
+Formatting options:
+
+- `t` – The name of the table to list.
+- `b` – The name of the base table (if inherited) of the table to list.
```
These are wrong. These options control the output formatting, what `-f t` means for example is that names of the _tables being listed_ should be displayed. Descriptions sound like it's a filter on what to list.

Compare for example with `dotnet info table Person` where `Person` is the name of the table to show info about.

per-samuelsson replied:
Example to illustrate:


```
class Foo { }
class Bar : Foo { }

star list table
Foo <none>
Bar Foo

star list table -f t
Foo
Bar

star list table -f b
<none>
Foo
```

per-samuelsson commented on docs/star-tool.md:
```diff
@@ -0,0 +1,226 @@
+- `t` – The name of the table to list.
+- `b` – The name of the base table (if inherited) of the table to list.
+
+To list all tables of an existing database located at `C:\databases\mydb`, including only user tables but both the name of the table and the name of its base table (if any), we run:
```
```suggestion
To list all tables of an existing database located at `C:\databases\mydb`, including only user tables and displaying both the name of the table and the name of its base table (if any), we run:
```

per-samuelsson commented on docs/star-tool.md:
````diff
@@ -0,0 +1,226 @@
+###### `list index`
+
+```
+star list index [options] <db> <table>
````
```suggestion
star list index [options] <db> [table]
```

Table here is optional. If no table is given, all indexes defined in the database (optionally filtered by `-us`) list listed.

erikvk replied:
Then I think we should start by changing the tool help printout, so it's correct, and then change the documentation to reflect it in a later PR. I think these docs should always mirror the help printouts.

per-samuelsson commented on docs/star-tool.md:
````diff
@@ -0,0 +1,226 @@
+star list index [options] <db> <table>
+```
+
+The `list index` sub-command lists all indexes in a table, the name of which is specified by the `<table>` argument, in the database referenced by the absolute or relative path given in the `<db>` argument. The following options are available to configure the output of the commmand:
````
This section has to be adapted according to above comment - `table` argument is optional.

per-samuelsson commented on docs/star-tool.md:
```diff
@@ -0,0 +1,226 @@
+
+Formatting options:
+
+- `i` – The name of the index to list.
```
Same applies here, as the comment I gave on `--format` on `list table`. It's the name of the current index being listed, not the one "to list".

erikvk replied:
I think we're reading these sentences differently. To me they mean exactly what you say they should mean. Basically [The name of] [the index to list] where we were about to list an index, and here we're deciding whether or not to include the name of it. Regardless, I'm changing these sentences :)

per-samuelsson commented on docs/star-tool.md:
````diff
@@ -0,0 +1,226 @@
+star info table [options] <db> <table>
+```
+
+The `info table` sub-command prints detailed information about a table, the name of which is specified in the `<table>` argument, in the database referenced by the absolute or relative path given in the `<db>` argument. The following options are available to configure the output of the commmand:
````
```suggestion
The `info table` sub-command prints detailed information about a table, the name of which is specified in the `<table>` argument, in the database referenced by the absolute or relative path given in the `<db>` argument. The following options are available to configure the output of the command:
```

per-samuelsson commented on docs/star-tool.md:
````diff
@@ -0,0 +1,226 @@
+star list table [options] <db>
+```
+
+The `list table` sub-command lists all tables in the database referenced by the absolute or relative path given in the `<db>` argument. The following options are available to configure the output of the commmand:
````
```suggestion
The `list table` sub-command lists all tables in the database referenced by the absolute or relative path given in the `<db>` argument. The following options are available to configure the output of the command:
```

per-samuelsson commented on docs/star-tool.md:
````diff
@@ -0,0 +1,226 @@
+star info index [options] <db> <index>
+```
+
+The `info index` sub-command prints detailed information about an index, the name of which is specified in the `<index>` argument, in the database referenced by the absolute or relative path given in the `<db>` argument. The following options are available to configure the output of the commmand:
````
```suggestion
The `info index` sub-command prints detailed information about an index, the name of which is specified in the `<index>` argument, in the database referenced by the absolute or relative path given in the `<db>` argument. The following options are available to configure the output of the command:
```

per-samuelsson reviewed (commented):
Second round review.

One addition could be to include the formatting defaults for each command supporting `--format` option.

Another addition could be to add sample output.

Commit:
Update docs/star-tool.md

Co-Authored-By: Per Samuelsson <per@starcounter.com>

Commit:
Update docs/star-tool.md

Co-Authored-By: Per Samuelsson <per@starcounter.com>

Commit:
Update docs/star-tool.md

Co-Authored-By: Per Samuelsson <per@starcounter.com>

erikvk committed:
Update docs/star-tool.md

Co-Authored-By: Per Samuelsson <per@starcounter.com>

erikvk committed:
Readability changes

erikvk committed:
Merge branch '3.0.0-rc-erikvk' of https://github.com/Starcounter/Docs into 3.0.0-rc-erikvk

erikvk commented:
@miyconst Let's merge this. I have addressed the review comments except https://github.com/Starcounter/Docs/pull/276#discussion_r359741428, since I think it should be fixed in the tool first.

Let's improve the docs further in future issues.
